### PR TITLE
Wh/optimization

### PR DESCRIPTION
**Pull request by whusted**
Integer and boolean constant folding implemented!
Please review and merge so we can be working on the same optimization spec.
- Still need to add float constant folding and will likely add handling trivial mathematical operations (x / 1, 1 \* x, etc.)

### CODE REVIEW

File: entities/block.coffee [hunk 1]
```diff
@@ -9,6 +9,7 @@ class Block
     statement.analyze(localContext) for statement in @statements
 
   optimize: ->
-    # TODO
-
+    @statements = (s.optimize() for s in @statements)
```
Comment (alexschneider) [inline entities/block.coffee]: Try using map and filter for this:

``` coffeescript
@statements = @statements.map((s) -> s.optimize()).filter((s) -> s?)
```

Comment (whusted) [inline entities/block.coffee]: For what reason?

Comment (rtoal) [inline entities/block.coffee]: Someone appears to me more fond of one-liners than _I_ am!

Comment (alexschneider) [inline entities/block.coffee]: Someone is :smile: 

It's not a huge deal, it just looked awkward to me.

Also, it's a nice [duck](https://rachelbythebay.com/w/2013/06/05/duck/).

Comment (whusted) [inline entities/block.coffee]: Ohhh I see. Haha alright I'll change it

**Pull request merged**